bindings/go/mio/mio.go

```diff
@@ -0,0 +1,490 @@
+    for _, pool := range pools {
+        if pool == "" {
+            return nil, nil // use default
+        }
+        id, err := ScanID(pool)
+        if err != nil {
+            return nil, fmt.Errorf("invalid pool: %v", pool)
+        }
+        res = new(C.struct_m0_fid)
+        *res = C.struct_m0_fid{id.u_hi, id.u_lo}
+        break
+    }
+    return res, nil
+}
+
+// Create creates object. Estimated object size must be specified
+// so that the optimal object unit size and block size for the best
+// I/O performance on the object could be calculated. Optionally,
+// the pool fid can be provided, if the object to be created on a
+// non-default pool.
+func (mio *Mio) Create(id string, sz uint64, pools ...string) error {
+    if mio.obj != nil {
+        return errors.New("object is already opened")
+    }
+    if err := mio.objNew(id); err != nil {
+        return err
+    }
+    pool, err := anyPool(pools)
+    if err != nil {
+        return err
+    }
+
+    lid, err := getOptimalUnitSz(sz)
+    if err != nil {
+        return fmt.Errorf("failed to figure out object unit size: %v", err)
+    }
+    C.m0_obj_init(mio.obj, &C.container.co_realm, &mio.objID, lid)
+
+    var op *C.struct_m0_op
+    rc := C.m0_entity_create(pool, &mio.obj.ob_entity, &op)
+    if rc != 0 {
+        return fmt.Errorf("failed to create object: %d", rc)
+    }
+    C.m0_op_launch(&op, 1)
+    rc = C.m0_op_wait(op, bits(C.M0_OS_FAILED,
+                               C.M0_OS_STABLE), C.M0_TIME_NEVER)
+    if rc == 0 {
+        rc = C.m0_rc(op)
+    }
+    C.m0_op_fini(op)
+    C.m0_op_free(op)
+
+    if rc != 0 {
+        return fmt.Errorf("create op failed: %d", rc)
+    }
+    mio.finishOpen(sz)
+
+    return nil
+}
+
+func roundupPower2(x int) (power int) {
+    for power = 1; power < x; power *= 2 {}
+    return power
+}
+
+const maxM0BufSz = 128 * 1024 * 1024
+
+// Estimate the optimal (block, group) sizes for the I/O
+func (mio *Mio) getOptimalBlockSz(bufSz int) (bsz, gsz int) {
+    if bufSz > maxM0BufSz {
+            bufSz = maxM0BufSz
+    }
+    pver := C.m0_pool_version_find(&C.instance.m0c_pools_common,
+                                   &mio.obj.ob_attr.oa_pver)
+    if pver == nil {
+            log.Panic("cannot find the object's pool version")
+    }
+    usz := int(C.m0_obj_layout_id_to_unit_size(C.ulong(mio.objLid)))
+    pa := &pver.pv_attr
+    gsz = usz * int(pa.pa_N)
+    /* max 2-times pool-width deep, otherwise we may get -E2BIG */
+    maxBs := int(C.uint(usz) * 2 * pa.pa_P * pa.pa_N / (pa.pa_N + 2 * pa.pa_K))
+
+    if bufSz >= maxBs {
+            return maxBs, gsz
+    } else if bufSz <= gsz {
+            return gsz, gsz
+    } else {
+            return roundupPower2(bufSz), gsz
+    }
+}
+
+func pointer2slice(p unsafe.Pointer, n int) []byte {
+    // Slice memory layout
+    var slice = struct {
+        addr uintptr
+        len  int
+        cap  int
+    }{uintptr(p), n, n}
```

Comment:
Hm.. this is quite low level. What about 
```
sice := (*reflect.SliceHeader)(p)
slice.Data = p
slice.Len = n
slice.Cap = n
```

Also, what prevents `slice` from being garbage collected? Don't you need a `runtime.KeepAlive()` somewhere?

Reply:
It was stolen from here https://github.com/golang/sys/blob/d75a52659825e75fff6158388dddc6a5b04f9ba5/unix/syscall_unix.go#L72-L80.

GC does not touch the memory that was not allocated by it (as far as I understand).

Reply:
But after you created a slice, how does Go run-time know that slice.addr came from C?

Reply:
I suppose it will know it from the fact that it was never allocated by Go. Let me quote the [doc](https://golang.org/cmd/cgo) here:

> Go is a garbage collected language, and the garbage collector needs to know the location of every pointer to Go memory. Because of this, there are restrictions on passing pointers between Go and C. 
>
> In this section the term Go pointer means a pointer to memory allocated by Go (such as by using the & operator or calling the predefined new function) and the term C pointer means a pointer to memory allocated by C (such as by a call to C.malloc). Whether a pointer is a Go pointer or a C pointer is a dynamic property determined by how the memory was allocated; it has nothing to do with the type of the pointer.

Reply:
Perfect.